### PR TITLE
Add `WorkerJobGroup` for serial tasks

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This adds a new type of job, a `WorkerJobGroup`, which allows you to define a group of tasks that run in order, but run in parallel with other groups. This lets you define tasks that are dependent upon each other to control how parallel things are.

Also allows individual `WorkerJob`s to depend on other jobs or `WorkerJobGroups`.

## Tests
 - Added a test to make sure tasks are executed in the correct order, and that no dependency loops are allowed.

### CODE REVIEW

File: lib/src/process_pool.dart [hunk 1]
```diff
@@ -7,14 +7,77 @@ import 'dart:convert' show Encoding;
...
+  /// Modifying the returned set will not affect the dependencies of this job.
+  ///
+  /// Will throw if there is a dependency cycle, or if the given job has not
+  /// been added to the pool.
+  ///
+  /// Defaults to an empty set.
+  Set<DependentJob> get dependsOn => _dependsOn.toSet();
+  final Set<DependentJob> _dependsOn;
+
+  void addDependency(DependentJob job) {
```
Comment (chunhtai) [inline lib/src/process_pool.dart:56]: These functions can use some doc

File: lib/src/process_pool.dart [hunk 2]
```diff
@@ -7,14 +7,77 @@ import 'dart:convert' show Encoding;
...
 
 import 'process_runner.dart';
 
+/// A type of job that can depend on other jobs.
+///
+/// This is the base type of [WorkerJob] and [WorkerJobGroup], which can all
+/// depend on each other.
+///
+/// Jobs are not allowed to depend on themselves, or have a dependency depend on
+/// them, even indirectly.
```
Comment (chunhtai) [inline lib/src/process_pool.dart]: `or have a dependency depend on them, even indirectly.`

Is there a Job class that does not inherited this class?

File: lib/src/process_pool.dart [hunk 3]
```diff
@@ -98,9 +164,65 @@ class WorkerJob {
...
+/// A job that groups other jobs.
+///
+/// Each worker job, except the first, are made to be dependent upon the
+/// previous worker job.
+///
+/// This group job is automatically dependent upon all of the [workers] it manages.
+///
+/// It can manage other groups, or individual [WorkerJob]s.
+class WorkerJobGroup extends DependentJob {
+  WorkerJobGroup(Iterable<DependentJob> workers,
```
Comment (chunhtai) [inline lib/src/process_pool.dart]: nit: maybe name this jobs or workerJobs?

File: lib/src/process_pool.dart [hunk 4]
```diff
@@ -98,9 +164,65 @@ class WorkerJob {
...
+      {Iterable<DependentJob>? dependsOn, this.name = 'Group'})
+      : assert(workers.isNotEmpty),
+        workers = <DependentJob>[...workers],
```
Comment (chunhtai) [inline lib/src/process_pool.dart]: nit: toList()

File: lib/src/process_pool.dart [hunk 5]
```diff
@@ -98,9 +164,65 @@ class WorkerJob {
...
+    jobs.add(this);
   }
+
+  @override
+  String toString() => name;
+}
+
+/// A job that groups other jobs.
+///
+/// Each worker job, except the first, are made to be dependent upon the
```
Comment (chunhtai) [inline lib/src/process_pool.dart]: Maybe also mention the jobs are run in order.

File: lib/src/process_pool.dart [hunk 6]
```diff
@@ -98,9 +164,65 @@ class WorkerJob {
...
+/// previous worker job.
+///
+/// This group job is automatically dependent upon all of the [workers] it manages.
```
Comment (chunhtai) [inline lib/src/process_pool.dart]: some of the doc here is more of implementation details, maybe just say this job finishes when all the worker finish.

File: lib/src/process_pool.dart [hunk 7]
```diff
@@ -235,9 +367,59 @@ class ProcessPool {
     return job;
   }
 
+  void _addFailedJob(WorkerJob job, ProcessRunnerException e) {
+    job.result = e.result ?? ProcessRunnerResult.failed;
+    job.exception = e;
+    _failedJobs.add(job);
+  }
+
+  DependentJob? _getNextInependentJob() {
```
Comment (chunhtai) [inline lib/src/process_pool.dart]:
```suggestion
  DependentJob? _getNextIndependentJob() {
```

File: lib/src/process_pool.dart (hunk 2 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart]: No, not at the moment.

File: lib/src/process_pool.dart (hunk 1 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart:56]: Done.

File: lib/src/process_pool.dart (hunk 5 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart]: Done.

File: lib/src/process_pool.dart (hunk 4 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart]: Done.

File: lib/src/process_pool.dart (hunk 3 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart]: Done.

File: lib/src/process_pool.dart (hunk 6 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart]: Done.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,11 @@
 # Change Log for `process_runner`
 
+## 4.2.0
+
+* Added `WorkerJobGroup` for running a group of dependent tasks in order.
```
Comment (chunhtai) [inline CHANGELOG.md]:
```suggestion
* Adds `WorkerJobGroup` for running a group of dependent tasks in order.
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,11 @@
...
+* Added the ability to have one job depend on another,
```
Comment (chunhtai) [inline CHANGELOG.md]:
```suggestion
* Adds the ability to have one job depend on another,
```

File: CHANGELOG.md (hunk 8 above)
Comment (chunhtai) [reply CHANGELOG.md]: use present tense, because this only happens when people upgrade. This is also more inline with first party package

File: analysis_options.yaml [hunk 10]
```diff
@@ -81,7 +81,6 @@ linter:
     - cancel_subscriptions
     # - cascade_invocations # not yet tested
     - cast_nullable_to_non_nullable
-    - collection_methods_unrelated_type
```
Comment (chunhtai) [inline analysis_options.yaml]: why is this removed? if this is a one of case, we should probably use // ignore ?

File: lib/src/process_pool.dart (hunk 2 above)
Comment (chunhtai) [reply lib/src/process_pool.dart]: In that case, should we remove the second sentence? I am not sure why Jobs are not allowed to have a dependency depend on them, or did you mean dependency cycle? If so, can you rephrase this sentence?

**Review by chunhtai — APPROVED** [review]
LGTM, only doc nitpicks

File: CHANGELOG.md (hunk 8 above)
Comment (gspencergoog) [reply CHANGELOG.md]: Done

File: CHANGELOG.md (hunk 9 above)
Comment (gspencergoog) [reply CHANGELOG.md]: Done

File: lib/src/process_pool.dart (hunk 2 above)
Comment (gspencergoog) [reply lib/src/process_pool.dart]: I meant a dependency cycle. Reworded.

File: analysis_options.yaml (hunk 10 above)
Comment (gspencergoog) [reply analysis_options.yaml]: Huh.  I removed it because I was getting an analyzer warning that it wasn't a valid lint anymore, but now that I put it back, there's no warning.
